Adding my next Custom Template - dataLayer Builder - Enhanced Ecommerce

Mikeulrich75 opened the pull request:
I have added my 2nd GTM custom template to your repository for you to review. As always, please let me know your thoughts and any suggestions to make this better. 

Thanks!

sahava commented on dataLayer Builder + Enhanced Ecommerce.tpl:
```diff
@@ -0,0 +1,554 @@
+    actionFieldParameters = actionFieldParameters;
+ } else {
+   actionFieldParameters = undefined;
+ }
```
This is all a bit elaborate when you could replace it with

```
const actionFieldParameterrs = data.actionFieldParameters ? makeTableMap(data.actionFieldParameters, 'name', 'value') : undefined;
```

Mikeulrich75 replied:
Oh weird, when I went into my custom template I had already switched this statement to shorthand. I must have given you an older version.

sahava commented on dataLayer Builder + Enhanced Ecommerce.tpl:
```diff
@@ -0,0 +1,554 @@
+  customParameter = customParameter;
+} else {
+  customParameter = undefined;
+}
```
Again, an elaborate way to set something as undefined. Just use:

```
const customParameter = data.customParameter ? makeTableMap(data.customParameter, 'name', 'value') : undefined;
```

Mikeulrich75 replied:
Same here!

sahava commented on dataLayer Builder + Enhanced Ecommerce.tpl:
```diff
@@ -0,0 +1,554 @@
+}
+//const customParameter = data.customParameter ? makeTableMap(data.customParameter, 'name', 'value') : {};
+const eventType = data.eventType;
+var dLayer = '';
```
As a coding pattern this is awkward, as you're intitializing `dLayer` as something it doesn't end up being (an object).

So

```
const dLayer = {};
```

and then instead of reinitializing it in the upcoming if...else blocks, just set its ecommerce property:

```
dLayer.ecommerce = {...};
```

Mikeulrich75 replied:
Good observation, the devils in the details. Thanks for the pointers on how to clean up my coding and how to code more efficiently. :)

sahava commented on dataLayer Builder + Enhanced Ecommerce.tpl:
```diff
@@ -0,0 +1,554 @@
+//const customParameter = data.customParameter ? makeTableMap(data.customParameter, 'name', 'value') : {};
+const eventType = data.eventType;
+var dLayer = '';
+if (eventType == 'impressions') {
```
No reason for `==` when you can be explicit with `===` :)

Mikeulrich75 replied:
Duly noted!

sahava commented on dataLayer Builder + Enhanced Ecommerce.tpl:
```diff
@@ -0,0 +1,554 @@
+const eventType = data.eventType;
+var dLayer = '';
+if (eventType == 'impressions') {
+  dLayer = {
```
```
dLayer.ecommerce = {...
```

Mikeulrich75 replied:
👍

sahava commented on dataLayer Builder + Enhanced Ecommerce.tpl:
```diff
@@ -0,0 +1,554 @@
+  dLayer = {
+  'ecommerce': {
+    'currencyCode': currencyParameter,
+    'actionField': actionFieldParameters,
```
`impressions` don't take an `actionField` parameter, nor does an `actionField` belong in the root `ecommerce` object.

Mikeulrich75 replied:
The actionField parameter has been blocked in the enabling conditions for Impressions, and also removed from the Impressions object in the script.

sahava commented on dataLayer Builder + Enhanced Ecommerce.tpl:
```diff
@@ -0,0 +1,554 @@
+  'ecommerce': {
+    'currencyCode': currencyParameter,
+    'promoView': { 
+     'actionField': actionFieldParameters, 
```
No `actionField` in promoView either :)

Mikeulrich75 replied:
I have removed the actionField parameter from the '**promoClick**' object in the custom template. I do however have an observation and question. I can't seem to set a 2nd enabling condition in the template to block for both '**Impressions**' and '**promoClick**' since there is no option to use RebEx, and it appears you can't have 2 enabling conditions in a row as they appear to cancel each other out.
I did try to create my own variable **const actionFieldEnable = ((eventType == ('impressions' || 'promoClick'))) ? 'false' : 'true';** and use this variable in the enabling conditions to block or fire based on a boolean but it didn't work either. It seems like custom templates is a bit limited without the use of RegEx or supporting the use of your own custom variables to enable fields. Do you have any suggestions on how to block enabling on multiple events?
 
Thanks in advance!

sahava replied:
Yeah, you can't use a "block" condition due to the thing you described (I've given feedback about this). 

You need to turn it around, and add all the "equals" options as possibilities.

So instead of someVar does not equal "one" and someVar does not equal "two", you have

someVar equals three
someVar equals four
someVar equals five

That will do until we get the regex.

Mikeulrich75 replied:
Thanks for the helpful tip, this did the trick, albeit very inefficiently! :) I really am hoping that Google will add in support for the RegEx object in the future, this will make all our lives a lot easier. 

Also, I did some very minor field name updates to make them a bit clearer. If you could do one more once over that would be great. This template should be good to go! 

Once more thing, apologies for not thinking about overwriting the main project readme, in future, I will make sure to put the whole think its own folder so this doesn't happen with future builds. 

Thank!

sahava commented:
Sorry about the delay in reviewing this!

Please check and double-check what you are trying to merge into the project. It looks to me like you are trying to add the files to the root of the project and thus end up overwriting the README.md of the project.

Make sure the tag ends up in the correct place! That would be /tags/utilities/EnhancedEcommerceDataLayerBuilder/

sahava commented:
@Mikeulrich75 your current PR is **still** making modifications to the README.md in the project root, and it is also trying to add both the Standard Ecommerce and Enhanced Ecommerce templates to the project root.

Create one more commit where you revert these changes. The PR should **only** include changes to the folder under tags/Utilities/EnhancedEcommerceDataLayerBuilder and nothing to the project root.

Mikeulrich75 commented:
Simo,

I am totally sorry about this, it was sloppy on my part, I should have double-checked that I had removed all of these files from the original PR before I committed. I have removed both the standard and enhanced ecommerce tpl files you referenced as well as my readme that was overwriting your project readme. In the future, I will be more careful. My team has just been overburdened lately with client work, so my attention has been heavily on client projects and having a hard time finding another analytics engineer (dev) to bring on to the team. But no excuses, thanks as always for your time. :)

sahava commented:
> Simo,
> 
> I am totally sorry about this, it was sloppy on my part, I should have double-checked that I had removed all of these files from the original PR before I committed. I have removed both the standard and enhanced ecommerce tpl files you referenced as well as my readme that was overwriting your project readme. In the future, I will be more careful. My team has just been overburdened lately with client work, so my attention has been heavily on client projects and having a hard time finding another analytics engineer (dev) to bring on to the team. But no excuses, thanks as always for your time. :)

Hey Michael, no worries :) Just one more thing - you are currently *deleting* the README.md file from the repo root, please submit a commit where the only files you change are the template and the readme for that template.

Mikeulrich75 commented:
Simo,

Thanks for the patience on this, it appears to be a bit tricky to merge updates to a PR in GitHub after the original has been submitted incorrectly. 

I have made sure that your README.md file is in my forked PR, and have made sure I only added my template (tpl) and README.md file into the appropriate folder. 

I hope everything has been done as you need to not overwrite anything, I know your probably tired of messing with this one. :)

If it isn't, would it just be easier to just close the original, and open a totally new one that has the correct structure? Just a thought. 

Also, just realized as I was reading your README.md file that the process must have changed from the 1st time I submitted a custom template, so I will make sure to always read the README.md file before submitting in the future, and follow explicitly. 

Thanks!
